client/cmd.go

```diff
@@ -148,31 +148,47 @@ func ReadTxCommandFlags(clientCtx Context, flagSet *pflag.FlagSet) (Context, err
 		return clientCtx, err
 	}
 
-	genOnly, _ := flagSet.GetBool(flags.FlagGenerateOnly)
-	clientCtx = clientCtx.WithGenerateOnly(genOnly)
+	if flagSet.Changed(flags.FlagGenerateOnly) {
```

Comment:
Because the tests are setting `GenerateOnly` to `true` on `ClientCtx` directly and this was overriding that.

Reply:
I would rather revert these changes to be honest. Otherwise, do you get the defaults?

Reply:
But aren't the defaults just `false`/empty string? Those defaults are already the ones on `client.Context` or am I missing something?

Reply:
No, I mean the default flag values. e.g. `bMode, _ := flagSet.GetString(flags.FlagBroadcastMode)`, the diff being that the current logic will populate `bMode` with the default flag value (`sync`), and with the new changes, it'll be `""`. Does that make sense?

Reply:
Ah, I see. The default that was setup when the flag was registered. We can revert this and change the tests to set the `--generate-only` flag, but it is confusing because in some cases `client.Context` can be pre-populated in tests and in other cases it will get overwritten. Not sure how to make this less clunky.

Reply:
The only default that seems to matter here is for `FlagBroadcastMode`. I would propose that we don't use these flags for defaults. Instead if there is an empty string in `client.Context.BroadcastMode` we should use the default of `sync` there. 

The problem with relying on flag defaults for `client.Context` defaults is that this causes unexpected behavior in integration tests when for instance someone writes `clientCtx.WithGenerateOnly(true)` but the actual behavior gets overwritten by `flagSet.GetBool(flags.FlagGenerateOnly)`. This was a bit tricky to debug and I don't want another developer to have to contend with the same thing. So my preference is to consistently use `flagSet.Changed`.

What do you think @alexanderbez ?

Reply:
Yeah, I can see how this is a bit confusing. So here is the gist:

- There are pre-defined "persistent" flags
- There are pre-defined "query" flags
- There are pre-defined "tx" flags

The only place where we check if a flag is explicitly set is for "persistent" flags (`ReadPersistentCommandFlags`) and that is because the `client.Context` may come pre-set with fields and we don't want to blindly override those. This was the original motivation for the `flagSet.Changed` calls. Everywhere else, we don't want to call  `flagSet.Changed` because we want the default flag values to propagate through.

However, as you've pointed out this is confusing and I was worried this would happen. 

----

Here is an approach that I think is simpler. We don't use `flagSet.Changed` _anywhere_ -- period. Instead, we always get the value of the flag IFF the respective `client.Context` field is empty/default/zero. This way, we can still pre-populate a `client.Context` and not blindly override, while also keeping the simple `flagSet.Get*` calls. WDYT?

Reply:
> Here is an approach that I think is simpler. We don't use `flagSet.Changed` _anywhere_ -- period. Instead, we always get the value of the flag IFF the respective `client.Context` field is empty/default/zero. This way, we can still pre-populate a `client.Context` and not blindly override, while also keeping the simple `flagSet.Get*` calls. WDYT?

That seems like a reasonable alternative, so instead of `flagSet.Changed` we'll have `if clientCtx.BroadcastMode == "" {`, etc. everywhere?

If we go with that approach I would suggest we leave it for a separate PR and unblock this one way or another.

Reply:
> That seems like a reasonable alternative, so instead of flagSet.Changed we'll have if clientCtx.BroadcastMode == "" {, etc. everywhere?

Close, not everywhere, but just in `ReadPersistentCommandFlags`, `ReadQueryCommandFlags`, and `ReadTxCommandFlags`. I'm OK with doing it in a separate PR. Can we at least undo the changes to this file?

Reply:
https://github.com/cosmos/cosmos-sdk/pull/6884